src/coreComponents/constitutive/fluid/BlackOilFluid.hpp

```diff
@@ -19,57 +19,1281 @@
 #ifndef GEOSX_CONSTITUTIVE_FLUID_BLACKOILFLUID_HPP_
 #define GEOSX_CONSTITUTIVE_FLUID_BLACKOILFLUID_HPP_
 
-#include "codingUtilities/EnumStrings.hpp"
-#include "constitutive/fluid/MultiFluidPVTPackageWrapper.hpp"
+#include "constitutive/fluid/BlackOilFluidBase.hpp"
+#include "constitutive/fluid/PVTOData.hpp"
+#include "functions/TableFunction.hpp"
+#include "codingUtilities/Utilities.hpp"
 
 namespace geosx
 {
 
 namespace constitutive
 {
 
-class BlackOilFluid : public MultiFluidPVTPackageWrapper
+static constexpr real64 minForPhasePresence = 1e-10;
+
+static constexpr localIndex NC_BO = 3;
+static constexpr localIndex NP_BO = 3;
+static constexpr localIndex HNC_BO = NC_BO - 1;
+
+/**
+ * @brief Kernel wrapper class for BlackOilFluid
+ *        This kernel can be called on the GPU
+ */
+class BlackOilFluidUpdate final : public MultiFluidBaseUpdate
 {
 public:
 
-  using exec_policy = serialPolicy;
+  BlackOilFluidUpdate( arrayView1d< integer const > const & phaseTypes,
```

Comment:
We could consider composition instead of inheritance for `Update` classes. I.e. `BlackOilFluidUpdate` would store an object of `MultiFluidBaseUpdate` and accept it in constructor. Thus all the members of that class (which make up 26 of the 36 parameters) would not need to be repeated here. We'd lose compiler-enforced virtual function signatures, but I feel more and more that we should get rid of `virtual` functions here anyway. We don't break these signatures often enough to justify `virtual` IMO.